synedrion/src/cggmp21/key_refresh.rs

```diff
@@ -1,321 +1,632 @@
+            _ => Err(MessageValidationError::InvalidEvidence("Invalid round number".into())),
+        }
+    }
+
+    fn verify_echo_broadcast_is_invalid(
+        deserializer: &Deserializer,
+        round_id: &RoundId,
+        message: &EchoBroadcast,
+    ) -> Result<(), MessageValidationError> {
+        match round_id {
+            r if r == &1 => message.verify_is_not::<Round1EchoBroadcast>(deserializer),
+            r if r == &2 => message.verify_is_not::<Round2EchoBroadcast<P, Id>>(deserializer),
+            r if r == &3 => message.verify_is_not::<Round3EchoBroadcast<Id>>(deserializer),
+            _ => Err(MessageValidationError::InvalidEvidence("Invalid round number".into())),
+        }
+    }
+
+    fn verify_normal_broadcast_is_invalid(
+        deserializer: &Deserializer,
+        round_id: &RoundId,
+        message: &NormalBroadcast,
+    ) -> Result<(), MessageValidationError> {
+        match round_id {
+            r if r == &1 => message.verify_is_some(),
+            r if r == &2 => message.verify_is_not::<Round2NormalBroadcast<P, Id>>(deserializer),
+            r if r == &3 => message.verify_is_not::<Round3Broadcast<P>>(deserializer),
+            _ => Err(MessageValidationError::InvalidEvidence("Invalid round number".into())),
+        }
+    }
 }
 
+/// Provable KeyRefresh faults.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 #[serde(bound(serialize = "
-    KeyRefreshErrorEnum<P>: Serialize,
+    Error<Id>: Serialize,
 "))]
 #[serde(bound(deserialize = "
-    KeyRefreshErrorEnum<P>: for<'x> Deserialize<'x>,
+    Error<Id>: for<'x> Deserialize<'x>,
 "))]
-pub struct KeyRefreshError<P: SchemeParams>(KeyRefreshErrorEnum<P>);
+pub struct KeyRefreshError<P, Id> {
+    error: Error<Id>,
+    phantom: PhantomData<P>,
+}
 
+impl<P, Id> From<Error<Id>> for KeyRefreshError<P, Id> {
+    fn from(source: Error<Id>) -> Self {
+        Self {
+            error: source,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<P, Id: PartyId> Display for KeyRefreshError<P, Id> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(
+            f,
+            "{}",
+            match self.error {
+                Error::R2HashMismatch => "Round 2: the previously sent hash does not match the public data.",
+                Error::R2WrongIdsX => "Round 2: wrong IDs in public shares map.",
+                Error::R2WrongIdsY => "Round 2: wrong IDs in Elgamal keys map.",
+                Error::R2WrongIdsA => "Round 2: wrong IDs in Schnorr commitments map.",
+                Error::R2PaillierModulusTooSmall => "Round 2: Paillier modulus is too small.",
+                Error::R2RPModulusTooSmall => "Round 2: ring-Pedersent modulus is too small.",
+                Error::R2NonZeroSumOfChanges => "Round 2: sum of share changes is not zero.",
+                Error::R2PrmFailed => "Round 2: `П^{prm}` verification failed.",
+                Error::R3ShareChangeMismatch { .. } =>
```

Comment:
Why not include the `reported_by` id in the message?

Reply:
We only have `Debug` guaranteed for IDs, which can be unwieldy (e.g. if it's a projective curve point, you'll get three huge coordinates), and `Display` is supposed to be short and tidy-looking. So I decided not to include the IDs there.

Reply:
In this specific case, `reported_by` doesn't really tell you much, it's not the guilty party, just a bit of an additional info to help with verification.